docker/webui/run_openqa.sh

```diff
@@ -1,15 +1,49 @@
 #!/bin/bash
 
-# prepare environment
-if [ ! -f /data/db/db.sqlite ]; then
-  /usr/share/openqa/script/initdb --user geekotest --init_database
-else
-  /usr/share/openqa/script/upgradedb --user geekotest --upgrade_database
-fi
+function waitForDBcreation() {
+  echo "Waiting for DB creation"
+  sleep 5
```

Comment:
Yes, this is a hack to solve this:

> Sometimes the docker-compose up fails because the web UIs try at the same time to create the DB
> 
> ```
> webui_2        | [2020-10-08 08:31:27.96315] [7] [warn] No configuration file supplied, will fallback to default configuration
> webui_2        | Can not create MOJO_TMPDIR : mkdir /var/lib/openqa/share/factory/tmp: Permission denied at /usr/share/openqa/script/../lib/OpenQA/Setup.pm line 288.
> webui_2        |
> webui_2        | failed to run SQL in /usr/share/openqa/script/../dbicdh/PostgreSQL/deploy/90/001-auto-__VERSION.sql: DBIx::Class::DeploymentHandler::DeployMethod::SQL::Translator::try {...} (): DBI Exception: DBD::Pg::db do failed: ERROR:  duplicate key value violates unique constraint "pg_type_typname_nsp_index"
> webui_2        | DETAIL:  Key (typname, typnamespace)=(dbix_class_deploymenthandler_versions_id_seq, 2200) already exists. at inline delegation in DBIx::Class::DeploymentHandler for deploy_method->deploy (attribute declared in /usr/lib/perl5/vendor_perl/5.26.1/DBIx/Class/DeploymentHandler/WithApplicatorDumple.pm at line 51) line 18
> webui_2        |  (running line 'CREATE TABLE dbix_class_deploymenthandler_versions ( id serial NOT NULL, version character varying(50) NOT NULL, ddl text, upgrade_sql text, PRIMARY KEY (id), CONSTRAINT dbix_class_deploymenthandler_versions_version UNIQUE (version) )') at /usr/lib/perl5/vendor_perl/5.26.1/DBIx/Class/DeploymentHandler/DeployMethod/SQL/Translator.pm line 263
> ```

I am not quite sure why this happens but with this wait I never saw an error like that. It doesn't happen if the DB is already created, but if not I think is a race condition. It could happens because the check that the DB exists and its creation is not atomic. 

An alternative to this solution could be to create a lock or semaphore but in this case I guess the code has to be inserted before and after the DB check and creation. 

Other solution can be:

```
if not DB exists then
  [tables].forEach
    create table
    if fails
      if not exists the DB table
         raise condition # because is an error
```

But please, I would be glad to hear a better solution.

Reply:
Resolved with the okurz idea